### PR TITLE
Block kit modals

### PR DESCRIPTION
**Pull request by jandonyada**
Block kit in modals docs

- [x] Optimized images
- [x] Grammarly check
- [x] Broken links check

- Added new doc /workbot/modals.md
- Added modals info in /workbot/block-kit.md
- Added modals info in New command trigger in /workbot/workbot-triggers.md
- Added new modal actions to /workbot/workbot-actions
- Added new modal actions to sidebar

### CODE REVIEW

File: docs/workbot/modals.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
+
+::: warning
+When a view is submitted, it closes by default. Use the correct View ID when updating views in response to view submissions.
+:::
+
+# Push modal view
+This action pushes a modal on top of the active view. To push a modal view, use **Trigger ID** (no View ID is required). You can find trigger ID from the datatree of the **New command** trigger.
+![Trigger ID](~@img/workbot/workbot-blockkit/trigger-id.png)
```
Comment (eeliang) [inline docs/workbot/modals.md]: \*Trigger ID for modal found in the New command trigger*

File: docs/workbot/modals.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+Typically, bot commands push views on top of the active view they exist in (**View ID**). The **View ID** can be found in the **New command trigger** datatree, under **Modals**.
+
+![Push modal](~@img/workbot/workbot-blockkit/push-modal.gif)
```
Comment (eeliang) [inline docs/workbot/modals.md]: \*Making changes on the active view*

File: docs/workbot/modals.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+| Root View ID     | View ID of the root view.                                                                                                                                                                                                     |
+| Previous View ID | View ID of the view beneath the current view                                                                                                                                                                                             |
+| Private metadata | Private data you can optionally use to pass to downstream recipes. This field is encrypted and hidden from users.                                                                                                                      |
+| Hash             | A unique value you can optionally use when updating modals. When provided, the hash is validated such that only the most recent view is updated, ensuring the correct view is being updated when updates are happening asynchronously. |
+
+# Modal view input
+The following table holds additional information about the **Modal** object and the data it correspondingly holds.
```
Comment (eeliang) [inline docs/workbot/modals.md]: The following table describes the configuration when working with Modals. This applies for the Open/update modal view action and the Push modal view action.

File: docs/workbot/modals.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+# Push modal view
+This action pushes a modal on top of the active view. To push a modal view, use **Trigger ID** (no View ID is required). You can find trigger ID from the datatree of the **New command** trigger.
+![Trigger ID](~@img/workbot/workbot-blockkit/trigger-id.png)
+
+Typically, bot commands push views on top of the active view they exist in (**View ID**). The **View ID** can be found in the **New command trigger** datatree, under **Modals**.
+
+![Push modal](~@img/workbot/workbot-blockkit/push-modal.gif)
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
```
Comment (eeliang) [inline docs/workbot/modals.md]: \*Root View ID and Previous View ID found in the New command trigger*

File: docs/workbot/workbot-actions.md [hunk 5]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
     </tr>
     </tbody>
 </table>
+
+## Modals
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: suggest to move
```
## Modals
Modals allow you to build rich, interactive and dynamic views that collect information from users in a structured manner.
![Modal example](~@img/workbot/workbot-blockkit/pretty-modal.png)
```

File: docs/workbot/workbot-actions.md [hunk 6]
```diff
@@ -13,6 +13,8 @@ Workbot supports 6 actions:
 * [Download attachment](#download-attachment)
 * [Return menu options](#return-menu-options)
 * [Upload file](#upload-file)
+* [Open/update modal view](#modals)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: \* \[open/update modal view\](\#open-update-modal-view)
\* \[push modal view\](\#push-modal-view)

File: docs/workbot/workbot-actions.md [hunk 7]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
     </tr>
     </tbody>
 </table>
+
+## Modals
+Modals allow you to build rich, interactive and dynamic views that collect information from users in a structured manner.
+![Modal example](~@img/workbot/workbot-blockkit/pretty-modal.png)
+
+### Modals input fields
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: suggest to remove
```
### Modals input fields
The following table holds additional information about the **Modal** object and the data it correspondingly holds.
<table>...</table>
```

File: docs/workbot/modals.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+Input blocks work differently from the interactive components from other blocks.
+
+While buttons, menus, and other interactive components invoke commands **when clicked**, input blocks only invoke commands **when a view is submitted**. For example, a user is free to choose from 'Approve' or 'Reject' – the value is only locked in once he submits the modal by clicking the submit button.
+
+![Example select input](~@img/workbot/workbot-blockkit/example-select-input.png)
+
+If a view does not have any input blocks, no submit button/close button needs to be defined. If you do have them, then you **must** define submit/close buttons.
+
+# Modal stack
+You can stack up to 3 views. Users see the top-most view. This view is called the active view.
```
Comment (eeliang) [inline docs/workbot/modals.md]: A modal can hold up to 3 views at one time in a view stack. Users **will only** see the top-most view, **also called** the active view. **View stacks are useful as it allows the user to return to previous views with their prior state still in place.**

File: docs/workbot/modals.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+![Modal stack with 3 views](~@img/workbot/workbot-blockkit/modal-stack.png)
+
+Opening a modal opens the root view. Pushing a modal view pushes a new view on top of any existing views.
```
Comment (eeliang) [inline docs/workbot/modals.md]: Opening a modal opens in the root view. **Updating a modal replaces the currently active view, while pushing a modal view applies a new view on top of the view stack.**

File: docs/workbot/modals.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+In contrast, modal submissions usually update the root view (Root View ID) or the previous view (Previous View ID). These views can be found in the **New command trigger** datatree, under **Modals**.
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
+
+::: warning
+When a view is submitted, it closes by default. Use the correct View ID when updating views in response to view submissions.
+:::
+
+# Push modal view
+This action pushes a modal on top of the active view. To push a modal view, use **Trigger ID** (no View ID is required). You can find trigger ID from the datatree of the **New command** trigger.
```
Comment (eeliang) [inline docs/workbot/modals.md]: This action requires an open modal view, use the \[previous action\]\(#opening-a-modal-view\) to open a model view.

File: docs/workbot/workbot-actions.md [hunk 11]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+In contrast, modal submissions usually update the root view (Root View ID) or the previous view (Previous View ID). These views can be found in the **New command trigger** datatree, under **Modals**.
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
+
+::: warning
+When a view is submitted, it closes by default. Use the correct View ID when updating views in response to view submissions.
+:::
+
+## Push modal view
+This action pushes a modal on top of the active view. To push a modal view, use **Trigger ID** (no View ID is required). You can find trigger ID from the datatree of the **New command** trigger.
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: This action requires an open modal view, use the \[previous action]\(\#opening-a-modal-view) to open a model view.

See modal input fields \[here\](/workbot/modals.md#modal-view-input)

File: docs/workbot/workbot-actions.md [hunk 12]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+![Trigger ID](~@img/workbot/workbot-blockkit/trigger-id.png)
+
+Typically, bot commands push views on top of the active view they exist in (**View ID**). The **View ID** can be found in the **New command trigger** datatree, under **Modals**.
+
+![Push modal](~@img/workbot/workbot-blockkit/push-modal.gif)
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: \*Root View ID and Previous View ID found in the New command trigger*

File: docs/workbot/workbot-actions.md [hunk 13]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+</table>
+
+## Open/update modal view
+You can open or update modals using the same action: Open/update modal view.
+![Open/update modal view](~@img/workbot/workbot-blockkit/open-update-modal-view.png)
+
+To open, update or push Modals, **Trigger ID** is always required. Trigger IDs are generated by Slack when users interact with:
+- buttons,
+- menus,
+- overflow
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: missing comma
\- overflow,

File: docs/workbot/workbot-actions.md [hunk 14]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+- modal view submissions
+
+### Opening a modal view
+Once a modal view is open, you can choose to update the view, or push a new view on top of it.
+
+To open a modal, use **Trigger ID**. You can find **Trigger ID** from the datatree of the **New command** trigger.
+![Trigger ID](~@img/workbot/workbot-blockkit/trigger-id.png)
+
+### Updating a modal view
+To update a modal, provide both the **Trigger ID** and the **View ID** of the view you want to update.
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: See modal input fields \[here\](/workbot/modals.md#modal-view-input)

File: docs/workbot/workbot-actions.md [hunk 15]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
+
+::: warning
+When a view is submitted, it closes by default. Use the correct View ID when updating views in response to view submissions.
+:::
+
+## Push modal view
+This action pushes a modal on top of the active view. To push a modal view, use **Trigger ID** (no View ID is required). You can find trigger ID from the datatree of the **New command** trigger.
+![Trigger ID](~@img/workbot/workbot-blockkit/trigger-id.png)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: \*Trigger ID for modal found in the New command trigger*

File: docs/workbot/workbot-actions.md [hunk 16]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+- shortcuts,
+- message actions,
+- slash commands,
+- modal view submissions
+
+### Opening a modal view
+Once a modal view is open, you can choose to update the view, or push a new view on top of it.
+
+To open a modal, use **Trigger ID**. You can find **Trigger ID** from the datatree of the **New command** trigger.
+![Trigger ID](~@img/workbot/workbot-blockkit/trigger-id.png)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: \*Trigger ID for modal found in the New command trigger*

File: docs/workbot/modals.md [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+## Updating a modal view
+To update a modal, provide both the **Trigger ID** and the **View ID** of the view you want to update.
+
+### When to update modals
+Typically, bot commands update the active view they exist in (View ID).
+![Update modal](~@img/workbot/workbot-blockkit/update-modal.gif)
+
+In contrast, modal submissions usually update the root view (Root View ID) or the previous view (Previous View ID). These views can be found in the **New command trigger** datatree, under **Modals**.
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
```
Comment (eeliang) [inline docs/workbot/modals.md]: \*Root View ID and Previous View ID found in the New command trigger*

File: docs/workbot/workbot-actions.md [hunk 18]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+
+### Updating a modal view
+To update a modal, provide both the **Trigger ID** and the **View ID** of the view you want to update.
+
+### When to update modals
+Typically, bot commands update the active view they exist in (View ID).
+![Update modal](~@img/workbot/workbot-blockkit/update-modal.gif)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: \*Making changes on the active view*

File: docs/workbot/workbot-actions.md [hunk 19]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+
+In contrast, modal submissions usually update the root view (Root View ID) or the previous view (Previous View ID). These views can be found in the **New command trigger** datatree, under **Modals**.
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: \*Root View ID and Previous View ID found in the New command trigger*

File: docs/workbot/workbot-actions.md [hunk 20]
```diff
@@ -305,3 +307,151 @@ The following table lists the fields available in an **Upload file** action.
...
+When a view is submitted, it closes by default. Use the correct View ID when updating views in response to view submissions.
+:::
+
+## Push modal view
+This action pushes a modal on top of the active view. To push a modal view, use **Trigger ID** (no View ID is required). You can find trigger ID from the datatree of the **New command** trigger.
+![Trigger ID](~@img/workbot/workbot-blockkit/trigger-id.png)
+
+Typically, bot commands push views on top of the active view they exist in (**View ID**). The **View ID** can be found in the **New command trigger** datatree, under **Modals**.
+
+![Push modal](~@img/workbot/workbot-blockkit/push-modal.gif)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: \*Making changes on the active view*

File: docs/workbot/workbot-triggers.md [hunk 21]
```diff
@@ -249,6 +249,15 @@ The following table holds additional information about the **Context** object an
     </tbody>
 </table>
 
+The following table holds additional information about the **Modal** object and the data it correspondingly holds.
```
Comment (eeliang) [inline docs/workbot/workbot-triggers.md]: \*\*Modal**s**\*\* object

**Comment by eeliang** [comment]
Also, I can't seem to open the ToC. This change fixes it for me.

```
      {
        title: 'Workbot actions',
        path: '/workbot/workbot-actions',
        children: [
          [
            '/workbot/workbot-actions#post-command-reply',
            'Post command reply'
          ],
          ['/workbot/workbot-actions#post-message', 'Post message'],
          [
            '/workbot/workbot-actions#post-notifications',
            'Post notifications'
          ],
          [
            '/workbot/workbot-actions#return-menu-options',
            'Return menu options'
          ],
          ['/workbot/workbot-actions#open-update-modal-view', 'Open/update modal view'],
          ['/workbot/workbot-actions#push-modal-view', 'Push modal view']
        ]
      },
```

File: docs/workbot/modals.md (hunk 8 above)
Comment (jandonyada) [reply docs/workbot/modals.md]: You worded it nicely, but the previous views may be returned to, but no remain at their prior state. Changing it to:

A modal can hold up to 3 views at one time in a view stack. Users will only see the top-most view, also called the active view. View stacks are useful as it allows the user to return to previous views, after submitting/closing their active view(s).

File: docs/workbot/modals.md [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+- Select menu input
+- Datepicker input
+- Checkboxes input
+
+# Interactive components vs input blocks
+Input blocks work differently from the interactive components from other blocks.
+
+While buttons, menus, and other interactive components invoke commands **when clicked**, input blocks only invoke commands **when a view is submitted**. For example, a user is free to choose from 'Approve' or 'Reject' – the value is only locked in once he submits the modal by clicking the submit button.
+
+![Example select input](~@img/workbot/workbot-blockkit/example-select-input.png)
```
[the remark this answers is not in the document]
Comment (jandonyada) [reply docs/workbot/modals.md]: What is your suggestion replacing here?

[the remark this answers is not in the document]
Comment (jandonyada) [reply docs/workbot/modals.md]: Or is it to add a caption after the image?

File: docs/workbot/modals.md [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+
+In contrast, modal submissions usually update the root view (Root View ID) or the previous view (Previous View ID). These views can be found in the **New command trigger** datatree, under **Modals**.
+
+![Modals object in New command trigger datatree](~@img/workbot/workbot-blockkit/modals-in-datatree.png)
+
+::: warning
+When a view is submitted, it closes by default. Use the correct View ID when updating views in response to view submissions.
+:::
+
+# Push modal view
```
Comment (jandonyada) [inline docs/workbot/modals.md]: @eeliang should this be # Push modal view action too for consistency?

File: docs/workbot/modals.md (hunk 10 above)
Comment (jandonyada) [reply docs/workbot/modals.md]: What should be replaced with your suggestion?

File: docs/workbot/workbot-actions.md (hunk 6 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: My worry is that if they skip to those sections, they will miss the modal input fields which is actually pretty important. Maybe we can put modal input fields in both sections? WDYT?

File: docs/workbot/modals.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (eeliang) [reply docs/workbot/modals.md]: It's to add a caption, sorry for being unclear

File: docs/workbot/workbot-actions.md (hunk 5 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: Not sure I understand here.

File: docs/workbot/workbot-actions.md (hunk 7 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: Not sure I understand here. Can you elaborate?

File: docs/workbot/modals.md (hunk 10 above)
Comment (eeliang) [reply docs/workbot/modals.md]: I think that this sentence should be added for clarity. That a user needs to open a modal before using the push modal view action

File: docs/workbot/workbot-actions.md (hunk 5 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: Where to move this?

File: docs/workbot/workbot-actions.md (hunk 14 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: Ah ok, so the anchors should go straight to the new actions, but in the new actions, point users to the input fields if they want to.

File: docs/workbot/workbot-actions.md (hunk 11 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: Do you mean to add this below the existing paragraph or replace?

File: docs/workbot/workbot-actions.md (hunk 6 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: Ok, after looking at your later comments, it makes more sense to direct users directly to the actions. And in those actions, point them to the modal input fields .

File: docs/workbot/modals.md (hunk 10 above)
Comment (jandonyada) [reply docs/workbot/modals.md]: Adding this as a new paragraph after this line.

File: docs/workbot/workbot-actions.md (hunk 11 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: Adding your suggestiong as a new paragraph after the original paragraph.

File: docs/workbot/workbot-actions.md [hunk 24]
```diff
@@ -13,6 +13,8 @@ Workbot supports 6 actions:
 * [Download attachment](#download-attachment)
 * [Return menu options](#return-menu-options)
 * [Upload file](#upload-file)
+* [Open/update modal view](#openupdate-modal-view-action)
```
Comment (eeliang) [inline docs/workbot/workbot-actions.md]: Can you check if this anchor works? Or needs to be changed to:
\* \[Open/update modal view]\(#open-update-modal-view-action)

File: docs/workbot/workbot-actions.md (hunk 13 above)
Comment (eeliang) [reply docs/workbot/workbot-actions.md]: Missed one

File: docs/workbot/workbot-actions.md (hunk 5 above)
Comment (eeliang) [reply docs/workbot/workbot-actions.md]: Sorry, typo. I meant "suggest to remove". But, on a second look. I think it is useful to keep a short description of Modals before moving to modal view actions.

File: docs/.vuepress/sidebar.js [hunk 25]
```diff
@@ -642,7 +643,8 @@ module.exports = [
             '/workbot/workbot-actions#return-menu-options',
             'Return menu options'
           ],
-          ['/workbot/workbot-actions#upload-file', 'Upload file']
+          ['/workbot/workbot-actions#modal-view-input', 'Open/update modal view'],
+          ['/workbot/workbot-actions#modal-view-input', 'Push modal view']
```
Comment (eeliang) [inline docs/.vuepress/sidebar.js]: Can we differentiate the anchors? Looks odd from the ToC
suggest:
- \(\#open-update-modal-view-action)
- \(\#push-model-view-action)

image

File: docs/workbot/workbot-actions.md (hunk 24 above)
Comment (jandonyada) [reply docs/workbot/workbot-actions.md]: yea I use a markdown previewer on Atom but I vuepress doesn't like it. #open-update-modal-view-action works

File: docs/.vuepress/sidebar.js (hunk 25 above)
Comment (jandonyada) [reply docs/.vuepress/sidebar.js]: Copy pasted wrongly

**Pull request merged**